File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (!node)
            return nullptr;
        unordered_map<Node*,Node*> mp;
        mp[node] = new Node(node->val);
        queue<Node*> q;
        q.push(node);
        while (!q.empty()){
            Node* cur = q.front();
            q.pop();
            for (Node* nb : cur->neighbors){
                if (!mp.contains(nb)){
                    mp[nb] = new Node(nb->val);
                    q.push(nb);
                }
                mp[cur]->neighbors.push_back(mp[nb]);
            }
        }
        return mp[node];
    }
};

File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node:
            return None
        mp = {node : Node(node.val)}
        q = deque()
        q.append(node)
        while len(q) != 0:
            cur = q.popleft()
            for nb in cur.neighbors:
                if not nb in mp:
                    mp[nb] = Node(nb.val)
                    q.append(nb)
                mp[cur].neighbors.append(mp[nb])
        return mp[node]